llvm/lib/Transforms/Scalar/SROA.cpp

```diff
@@ -3313,11 +3315,24 @@ class llvm::sroa::AllocaSliceRewriter
         SrcPtr = OurPtr;
         SrcAlign = SliceAlign;
       }
-      // XXX: copying the II.shouldPreserveCheriTags() flags may be overly
-      // restrictive for optimization purposes.
-      CallInst *New =
-          IRB.CreateMemCpy(DestPtr, DestAlign, SrcPtr, SrcAlign, Size,
-                           II.shouldPreserveCheriTags(), II.isVolatile());
+
+      auto PreserveTags = II.shouldPreserveCheriTags();
+      if (PreserveTags != PreserveCheriTags::Unnecessary &&
+          !canRangeContainCapabilities(getCapabilitySize(DL), NewBeginOffset,
+                                       NewEndOffset)) {
+        // If we see a new slice that is less than the capability size, we can
+        // set the memcpy() flag to PreserveCheriTags::Unnecessary even if the
+        // whole struct copy had PreserveCheriTags::Required.
+        // TODO: Copying PreserveCheriTags::Required for any slice >= capability
+        // size may be overly restrictive. Should we just drop the attribute
+        // unless the value is PreserveCheriTags::Unnecessary? The resulting
+        // code should still be correct since the backend needs to be
```

Comment:
I believe adding must_preserve_tags was a mistake and we should eventually drop it. Optimizations are not required to preserve attributes so it's rather fragile and the backends have to be conservative anyway if there isn't an attribute.